compiler/src/iree/compiler/Codegen/Common/FoldAffineMinWithSCFIf.cpp

```diff
@@ -0,0 +1,199 @@
+  if (minOp.getNumOperands() != 1) return failure();
+
+  AffineMap map = minOp.getAffineMap();
+  if (map.getNumResults() != 2) return failure();
+
+  OpFoldResult lbOfr, ubOfr, tileSizeOfr;
+  if (failed(matchTileAndDistributionLoop(minOp.getOperand(0), lbOfr, ubOfr,
+                                          tileSizeOfr)))
+    return failure();
+
+  Optional<int64_t> lbOpt = getConstantIntValue(lbOfr);
+  if (!lbOpt) return failure();
+  Optional<int64_t> ubOpt = getConstantIntValue(ubOfr);
+  if (!ubOpt) return failure();
+  Optional<int64_t> tileSizeOpt = getConstantIntValue(tileSizeOfr);
+  if (!tileSizeOpt) return failure();
+
+  lb = *lbOpt;
+  ub = *ubOpt;
+  tileSize = *tileSizeOpt;
+  return success();
+}
+
+/// Folds `affine.min` if its constant value argument is used by scf.if,
+/// where the op is folded into one of the two values.
+///
+/// ```mlir
+/// %min = affine.min(%a, %cst)
+/// %cmp = icmp.eq %min, %cst
+/// scf.if %cmp {
+///   // expect the uses of %min are already replaced by %cst
+/// } else {
+///   // use of %min can be replaced by %a
+/// }
+/// ```
+struct FoldAffineMinWithSCFIf final : public OpRewritePattern<scf::IfOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::IfOp ifOp,
+                                PatternRewriter &rewriter) const override {
+    auto cmpOp = ifOp.getCondition().getDefiningOp<arith::CmpIOp>();
+    if (!cmpOp) return failure();
+    if (cmpOp.getPredicate() != arith::CmpIPredicate::eq) return failure();
+
+    AffineMinOp minOp;
+    Value b;
+    if (isa<AffineMinOp>(cmpOp.getLhs().getDefiningOp())) {
+      minOp = cast<AffineMinOp>(cmpOp.getLhs().getDefiningOp());
+      b = cmpOp.getRhs();
+    } else if (isa<AffineMinOp>(cmpOp.getRhs().getDefiningOp())) {
+      minOp = cast<AffineMinOp>(cmpOp.getRhs().getDefiningOp());
+      b = cmpOp.getLhs();
+    } else {
+      return failure();
+    }
+    auto cstOp = dyn_cast<arith::ConstantIndexOp>(b.getDefiningOp());
+    if (!cstOp) return failure();
+    int64_t cstInt = cstOp.value();
+
+    AffineMap minMap = minOp.getAffineMap();
+    if (minMap.getNumResults() != 2) return failure();
+
+    FailureOr<unsigned> affineCstIndexOrFailure =
+        findIndexOfConstResult(minMap, cstInt);
+    if (failed(affineCstIndexOrFailure)) return failure();
+    unsigned affineCstIndex = *affineCstIndexOrFailure;
+
+    // find the uses of min in the else block.
+    SmallVector<OpOperand *> usesInElse =
+        getUsesInRegion(minOp, &ifOp.getElseRegion());
+    if (usesInElse.empty()) return failure();
+
+    OpBuilder builder(ifOp);
+    OpBuilder::InsertionGuard guard(builder);
+    int64_t lb, ub, tileSize;
+    Value valueToReplace;
+    if (succeeded(matchBoundedTileSizeOp(minOp, lb, ub, tileSize))) {
+      // The other part is the partial tile size, (ub - lb) % tileSize.
+      valueToReplace = builder.create<arith::ConstantIndexOp>(
+          minOp.getLoc(), (ub - lb) % tileSize);
+    } else {
```

Comment:
so those math are true only if the tile size is the same as the cstInt above? and the affine map has to be ub - iv? Is this checked anywhere?